## **HW4: Feature Engineering and Linear Regression**

### **TODO:Ramsay Ward

**Attention:** This is an individual assignment.

#### **Description of the Dataset**

* The Ames Housing dataset describes the sale of individual residential properties in Ames, Iowa, from 2006 to 2010. It contains a wide range of features, making it an excellent dataset to practice feature engineering techniques. In this assignment, you will explore a series of feature engineering tasks aimed at improving linear regression predictions.

* Please make sure you have read the  provided`data_description.txt` file that provides additional information about the dataset and its features before you start implementing your homework.

### **Question 1: Import libraries, Load Train and Test datasets into separate DataFrames**

In [1]:
# Write your code here
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score
import matplotlib.pyplot as plt


# Load train and test datasets
train = pd.read_csv('train.csv')
test = pd.read_csv('test.csv')

train.head()



Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities  ... PoolArea PoolQC Fence MiscFeature MiscVal MoSold  \
0         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
1         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      5   
2         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      9   
3         Lvl    AllPub  ...        0    NaN   NaN         NaN       0      2   
4         Lvl    AllPub  ...        0    NaN   NaN         NaN       0     12   

  YrSold  SaleType  SaleCondition  SalePrice  
0   2008        WD         Normal     208500  
1   2007        WD         Normal     181500  
2   2008        WD         Normal     223500  
3   2006        WD        Abnorml     140000  
4   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [2]:
# Optional diplay all columns to be able to see all the columns (features)

pd.set_option('display.max_columns', None)
train.columns

Index(['Id', 'MSSubClass', 'MSZoning', 'LotFrontage', 'LotArea', 'Street',
       'Alley', 'LotShape', 'LandContour', 'Utilities', 'LotConfig',
       'LandSlope', 'Neighborhood', 'Condition1', 'Condition2', 'BldgType',
       'HouseStyle', 'OverallQual', 'OverallCond', 'YearBuilt', 'YearRemodAdd',
       'RoofStyle', 'RoofMatl', 'Exterior1st', 'Exterior2nd', 'MasVnrType',
       'MasVnrArea', 'ExterQual', 'ExterCond', 'Foundation', 'BsmtQual',
       'BsmtCond', 'BsmtExposure', 'BsmtFinType1', 'BsmtFinSF1',
       'BsmtFinType2', 'BsmtFinSF2', 'BsmtUnfSF', 'TotalBsmtSF', 'Heating',
       'HeatingQC', 'CentralAir', 'Electrical', '1stFlrSF', '2ndFlrSF',
       'LowQualFinSF', 'GrLivArea', 'BsmtFullBath', 'BsmtHalfBath', 'FullBath',
       'HalfBath', 'BedroomAbvGr', 'KitchenAbvGr', 'KitchenQual',
       'TotRmsAbvGrd', 'Functional', 'Fireplaces', 'FireplaceQu', 'GarageType',
       'GarageYrBlt', 'GarageFinish', 'GarageCars', 'GarageArea', 'GarageQual',
       'GarageCond', 'PavedDrive

### **Question 2: Identify columns with missing data and then determine an appropriate strategy for each.**

In [14]:
# Write your code here

train_missing_data = train.isnull().sum()
train_missing_data = train_missing_data[train_missing_data > 0 ]
train_missing_data = train_missing_data.sort_values(ascending=False)

test_missing_data = test.isnull().sum()
test_missing_data = test_missing_data[test_missing_data > 0 ]
test_missing_data = test_missing_data.sort_values(ascending = False)

print('missing train data', train_missing_data)
print('missing test data', test_missing_data )

train.drop(train_missing_data)
test.drop(test_missing_data)


missing train data PoolQC          1453
MiscFeature     1406
Alley           1369
Fence           1179
FireplaceQu      690
LotFrontage      259
GarageType        81
GarageYrBlt       81
GarageFinish      81
GarageQual        81
GarageCond        81
BsmtExposure      38
BsmtFinType2      38
BsmtFinType1      37
BsmtCond          37
BsmtQual          37
MasVnrArea         8
MasVnrType         8
Electrical         1
dtype: int64
missing test data PoolQC          1456
MiscFeature     1408
Alley           1352
Fence           1169
FireplaceQu      730
LotFrontage      227
GarageCond        78
GarageYrBlt       78
GarageQual        78
GarageFinish      78
GarageType        76
BsmtCond          45
BsmtExposure      44
BsmtQual          44
BsmtFinType1      42
BsmtFinType2      42
MasVnrType        16
MasVnrArea        15
MSZoning           4
BsmtFullBath       2
BsmtHalfBath       2
Functional         2
Utilities          2
GarageCars         1
GarageArea         1
TotalBsmtSF        1
Kitch

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0     1461          20       RH         80.0    11622   Pave   NaN      Reg   
3     1464          60       RL         78.0     9978   Pave   NaN      IR1   
5     1466          60       RL         75.0    10000   Pave   NaN      IR1   
6     1467          20       RL          NaN     7980   Pave   NaN      IR1   
7     1468          60       RL         63.0     8402   Pave   NaN      IR1   
...    ...         ...      ...          ...      ...    ...   ...      ...   
1453  2914         160       RM         21.0     1526   Pave   NaN      Reg   
1454  2915         160       RM         21.0     1936   Pave   NaN      Reg   
1455  2916         160       RM         21.0     1894   Pave   NaN      Reg   
1457  2918          85       RL         62.0    10441   Pave   NaN      Reg   
1458  2919          60       RL         74.0     9627   Pave   NaN      Reg   

     LandContour Utilities LotConfig LandSlope Neighborhood Condition1  \
0            Lvl    AllPub    Inside       Gtl        NAmes      Feedr   
3            Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
5            Lvl    AllPub    Corner       Gtl      Gilbert       Norm   
6            Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
7            Lvl    AllPub    Inside       Gtl      Gilbert       Norm   
...          ...       ...       ...       ...          ...        ...   
1453         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
1454         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
1455         Lvl    AllPub    Inside       Gtl      MeadowV       Norm   
1457         Lvl    AllPub    Inside       Gtl      Mitchel       Norm   
1458         Lvl    AllPub    Inside       Mod      Mitchel       Norm   

     Condition2 BldgType HouseStyle  OverallQual  OverallCond  YearBuilt  \
0          Norm     1Fam     1Story            5            6       1961   
3          Norm     1Fam     2Story            6            6       1998   
5          Norm     1Fam     2Story            6            5       1993   
6          Norm     1Fam     1Story            6            7       1992   
7          Norm     1Fam     2Story            6            5       1998   
...         ...      ...        ...          ...          ...        ...   
1453       Norm    Twnhs     2Story            4            5       1970   
1454       Norm    Twnhs     2Story            4            7       1970   
1455       Norm   TwnhsE     2Story            4            5       1970   
1457       Norm     1Fam     SFoyer            5            5       1992   
1458       Norm     1Fam     2Story            7            5       1993   

      YearRemodAdd RoofStyle RoofMatl Exterior1st Exterior2nd MasVnrType  \
0             1961     Gable  CompShg     VinylSd     VinylSd       None   
3             1998     Gable  CompShg     VinylSd     VinylSd    BrkFace   
5             1994     Gable  CompShg     HdBoard     HdBoard       None   
6             2007     Gable  CompShg     HdBoard     HdBoard       None   
7             1998     Gable  CompShg     VinylSd     VinylSd       None   
...            ...       ...      ...         ...         ...        ...   
1453          1970     Gable  CompShg     CemntBd     CmentBd       None   
1454          1970     Gable  CompShg     CemntBd     CmentBd       None   
1455          1970     Gable  CompShg     CemntBd     CmentBd       None   
1457          1992     Gable  CompShg     HdBoard     Wd Shng       None   
1458          1994     Gable  CompShg     HdBoard     HdBoard    BrkFace   

      MasVnrArea ExterQual ExterCond Foundation BsmtQual BsmtCond  \
0            0.0        TA        TA     CBlock       TA       TA   
3           20.0        TA        TA      PConc       TA       TA   
5            0.0        TA        TA      PConc       Gd       TA   
6            0.0        TA        Gd      PConc       Gd       TA   

### **Question 3: One-hot Encoding for Categorical Variables**

* Apply one-hot encoding to transform columns with categorical values.

In [15]:
# Write your code here
cat_columns = train.select_dtypes(include= ['object']).columns.tolist()
train_encoded = pd.get_dummies(train, columns= cat_columns)
train_encoded.head()

#test_encoded = pd.get_dummies(test, columns= cat_columns)
#test_encoded.head()

Id  MSSubClass  LotFrontage  LotArea  OverallQual  OverallCond  YearBuilt  \
0   1          60         65.0     8450            7            5       2003   
1   2          20         80.0     9600            6            8       1976   
2   3          60         68.0    11250            7            5       2001   
3   4          70         60.0     9550            7            5       1915   
4   5          60         84.0    14260            8            5       2000   

   YearRemodAdd  MasVnrArea  ExterQual  ExterCond  BsmtFinSF1  BsmtFinSF2  \
0          2003       196.0          4          3         706           0   
1          1976         0.0          3          3         978           0   
2          2002       162.0          4          3         486           0   
3          1970         0.0          3          3         216           0   
4          2000       350.0          4          3         655           0   

   BsmtUnfSF  TotalBsmtSF  1stFlrSF  2ndFlrSF  LowQualFinSF  GrLivArea  \
0        150          856       856       854             0       1710   
1        284         1262      1262         0             0       1262   
2        434          920       920       866             0       1786   
3        540          756       961       756             0       1717   
4        490         1145      1145      1053             0       2198   

   BsmtFullBath  BsmtHalfBath  FullBath  HalfBath  BedroomAbvGr  KitchenAbvGr  \
0             1             0         2         1             3             1   
1             0             1         2         0             3             1   
2             1             0         2         1             3             1   
3             1             0         1         0             3             1   
4             1             0         2         1             4             1   

   TotRmsAbvGrd  Fireplaces  GarageYrBlt  GarageCars  GarageArea  WoodDeckSF  \
0             8           0       2003.0           2         548           0   
1             6           1       1976.0           2         460         298   
2             6           1       2001.0           2         608           0   
3             7           1       1998.0           3         642           0   
4             9           1       2000.0           3         836         192   

   OpenPorchSF  EnclosedPorch  3SsnPorch  ScreenPorch  PoolArea  MiscVal  \
0           61              0          0            0         0        0   
1            0              0          0            0         0        0   
2           42              0          0            0         0        0   
3           35            272          0            0         0        0   
4           84              0          0            0         0        0   

   MoSold  YrSold  SalePrice  TotalArea  LotAreaBin  LotAreaQuantileBin  \
0       2    2008     208500       2566           0                   1   
1       5    2007     181500       2524           0                   2   
2       9    2008     223500       2706           0                   3   
3       2    2006     140000       2473           0                   2   
4      12    2008     250000       3343           0                   4   

   MSZoning_C (all)  MSZoning_FV  MSZoning_RH  MSZoning_RL  MSZoning_RM  \
0                 0            0            0            1            0   
1                 0            0            0            1            0   
2                 0            0            0            1            0   
3                 0            0            0            1            0   
4                 0            0            0            1            0   

   Street_Grvl  Street_Pave  Alley_Grvl  Alley_Pave  LotShape_IR1  \
0            0            1           0           0             0   
1            0            1           0           0             0   
2            0            1           0           0             1   
3          

### **Question 4: Scaling and Normalization**

* Using `StandarScaler` from `scikit-learn`, scale the features in some of the numeric columns.
* You will need to import required libraries and modules here as well.

In [5]:
# Write your code here
import numpy as np
from sklearn.preprocessing import StandardScaler


# Identify numeric columns and put them in a python list
numeric_columns = train.select_dtypes(include= [np.number]).columns

# From the `numeric_columns`, exclude the target column 'SalePrice' and 'Id' column
numeric_columns = numeric_columns.drop(['SalePrice', 'Id'])
numeric_columns.tolist()

# Scaling the numeric columns, initialize StabdardScaler from `scikit-learn`
scaler = StandardScaler()

# fit and transform the numeric columns using the scaler you defined above
train_encoded[numeric_columns] = scaler.fit_transform(train[numeric_columns])

### **Question 5: Feature Extraction from Year Variables**

* Extract the information about the age of the house at the time of the sale by calculating the difference between `YrSold` and `YearBuilt` features and create a new Column (feature) named `HouseAge` from the result.
* Extract the number of years since remodelling when the house was sold by calculating the difference between `YrSold` and `YearRemodAdd` features and create a new Column (feature) named `YearsSinceRemod` from the result.

In [6]:
# Write your code here

# Age of the house at the time of sale
train_encoded['HouseAge'] = train_encoded['YrSold'] - train_encoded['YearBuilt']

# Number of years since remodelling when the house was sold
train_encoded['YearsSinceRemod'] = train_encoded['YrSold'] - train_encoded['YearRemodAdd']

# Display the new features you have created above
train_encoded[['HouseAge', 'YearsSinceRemod']].head()

HouseAge  YearsSinceRemod
0 -0.912216        -0.739891
1 -0.771172        -0.184862
2 -0.845975        -0.691437
3  0.495977        -0.647357
4 -0.812854        -0.594530

### **Question 6: Dimensionality Reduction**

* In this question you will apply the PCA algorithm to reduce the dimensions of the dataset. PCA will not work if there are missing values in your data.For this, you will first fill all the missing values in the numerical columns with the median of the column.
* Then, you will fill the missing categorical columns with mode imputation.
* Tip: To fill all the missing values in all columns, first identify all numeric columns with missing data and store them in a variable called `numeric_cols_with_missing`. Do the same for categoical columns with missing data and store them into a variable called `categorical_cols_with_missing`. Then, you can iterate over (using a for loop) these two sets of columns and fill them with the appropriate values described above. Do not forget to use `inplace=True` attribute in the `fillna()` method that you will use to fill missing values so that you manipulate the original dataset without creating a copy of it.

In [16]:
# Write your code here
# Numeric columns: median imputation
numeric_col = train_encoded.select_dtypes(include= ['int64','float64']).columns.tolist()

numeric_cols_with_missing = train_encoded[numeric_col].isnull().sum()


for col in numeric_cols_with_missing.index:
    # fill the missing columns
    train_encoded[col].fillna(train_encoded[col].median(), inplace = True)


# Categorical columns: mode imputation
cat_col = train_encoded.select_dtypes(include= ['object']).columns.tolist()

categorical_cols_with_missing = train_encoded[cat_col].isnull().sum()

for col in categorical_cols_with_missing.index:
    # fill the missing columns
    train_encoded[col].fillna(train_encoded[col].mode().iloc[0], inplace = True)

# Verifying that there are no more missing values
assert train_encoded.isnull().sum().sum() == 0

#### **Implement the PCA algorithm to the dataset with no missing values**


**IMPORTANT:**  Load the original `train.csv` again for this question.

To complete this question, please follow the instructions listed below. Keep in mind that this is a general guide and serves as a foundational checklist, and based on your specific methodology, additional or alternative steps may be necessary:

* Load the dataset again
* Seperate the target variable
* Drop any column that does not have any meaning in this context ( i.e Id)
* Identify the columns with missing values and calculate their percentages
* Drop columns with more than 80% missing values
* Impute numerical columns with their median
* Impute categorical columns with their mode
* Encode categorical columns using one hot encoding
* Scale the dataset so that each feature has a mean of 0 and a standard deviation of 1
* Apply PCA and determine the number of components for 95% variance
* Apply PCA `n_components_` method to retain approximately 95% of the variance


In [17]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler
from sklearn.decomposition import PCA

# Write your code here
df_train = pd.read_csv('train.csv')

target = df_train['SalePrice']
df_train.drop(columns = ['Id','SalePrice'])

missing_percent = df_train.isnull().mean()

drop_col = missing_percent[missing_percent > .80].index
df_train.drop(columns = drop_col, inplace = True)

num_col = df_train.select_dtypes(include = [np.number]).columns
df_train[num_col] = df_train[num_col].fillna(df_train[num_col].median())

cat_col = df_train.select_dtypes(include = ['object']).columns
df_train[cat_col] = df_train[cat_col].fillna(df_train[cat_col].mode().iloc[0])

train_encoded = pd.get_dummies(df_train)

scaler =StandardScaler()
train_scaled = scaler.fit_transform(train_encoded)

pca = PCA(n_components = 0.95)
train_pca = pca.fit_transform(train_scaled)

# Number of components retained after PCA
num_components = pca.n_components_
num_components

174

### **Question 7: Engineering Ordinal Features**

* For ordinal columns like `ExterQual` and `ExterCond`,  map the values to numbers (using the `map()`(https://docs.python.org/3/library/functions.html#map) method.

In [9]:
# Write your code here
# Mapping ordinal values to numbers
ordinal_mappings = {
    'Ex': 5,
    'Gd': 4,
    'TA': 3,
    'Fa': 2,
    'Po': 1
}

train['ExterQual'] = train['ExterQual'].map(ordinal_mappings)
train['ExterCond'] = train['ExterCond'].map(ordinal_mappings)


train[['ExterQual', 'ExterCond']].head()

ExterQual  ExterCond
0          4          3
1          3          3
2          4          3
3          3          3
4          4          3

### **Question 8: Feature Interaction**

* An **interaction feature** refers to a new feature that is created by combining or relating two or more existing features. It is based on the idea that two or more variables together may have a synergistic effect on the target variable that is not captured when they are used independently.

* Create an interaction feature, such as the total area of the house by using the available features `ToralBsmtSF`, `1stFlrSF`, and `2ndFlrSF`
* Tip: You will need to add these `Series` to get the `TotalArea`

In [10]:
# Write your code here

# Total area of the house
train['TotalArea'] = train['TotalBsmtSF'] + train['1stFlrSF'] + train['2ndFlrSF']

# Displaying the new feature
train[['TotalArea']].head()

TotalArea
0       2566
1       2524
2       2706
3       2473
4       3343

### **Question 9: Binning**

* Group the `LotArea` feature into 5 bins

In [11]:
# Write your code here

train['LotAreaBin'] = pd.cut(train['LotArea'], bins = 5, labels = False)

# Displaying the binned column
train[['LotArea', 'LotAreaBin']].head()

LotArea  LotAreaBin
0     8450           0
1     9600           0
2    11250           0
3     9550           0
4    14260           0

In [12]:
# Check the unique values in the 'LotAreaBin' column and the corresponding bin boundaries
bins = pd.cut(train['LotArea'], bins = 5).unique()
unique_values = train['LotAreaBin'].value_counts().sort_index()

bins, unique_values

([(1086.055, 44089.0], (44089.0, 86878.0], (129667.0, 172456.0], (172456.0, 215245.0], (86878.0, 129667.0]]
 Categories (5, interval[float64, right]): [(1086.055, 44089.0] < (44089.0, 86878.0] <
                                            (86878.0, 129667.0] < (129667.0, 172456.0] <
                                            (172456.0, 215245.0]],
 0    1447
 1       9
 2       1
 3       2
 4       1
 Name: LotAreaBin, dtype: int64)

* From the results, it's evident that the vast majority of houses (1447 out of 1460) have lot areas that fall into the first bin (Bin 0), which corresponds to the interval (1086.055,44089.0 (1086.055,44089.0]. This is why you see many values of 0 in the `LotAreaBin` column.

* The binning behavior here is due to a few properties with very large lot areas that are influencing the range and hence the bin boundaries. The first bin captures most of the data, while the other bins capture only a few outliers. In such cases, it might be more appropriate to use quantile-based binning (using pd.qcut()) to ensure a more even distribution of data points across bins.

In [13]:
# Observe the results of the quantile-based binning and compare it to the result from only `pd.cut()` above

train['LotAreaQuantileBin'] = pd.qcut(train['LotArea'], q=5, labels=False)

# Check the unique values in the 'LotAreaQuantileBin' column and the corresponding bin boundaries
quantile_bins = pd.qcut(train['LotArea'], q=5, retbins=True)
unique_values_quantile = train['LotAreaQuantileBin'].value_counts().sort_index()

quantile_bins, unique_values_quantile

((0          (7078.4, 8793.4]
  1         (8793.4, 10198.2]
  2        (10198.2, 12205.8]
  3         (8793.4, 10198.2]
  4       (12205.8, 215245.0]
                 ...         
  1455       (7078.4, 8793.4]
  1456    (12205.8, 215245.0]
  1457      (8793.4, 10198.2]
  1458      (8793.4, 10198.2]
  1459      (8793.4, 10198.2]
  Name: LotArea, Length: 1460, dtype: category
  Categories (5, interval[float64, right]): [(1299.999, 7078.4] < (7078.4, 8793.4] < (8793.4, 10198.2] <
                                             (10198.2, 12205.8] < (12205.8, 215245.0]],
  array([  1300. ,   7078.4,   8793.4,  10198.2,  12205.8, 215245. ])),
 0    292
 1    292
 2    292
 3    292
 4    292
 Name: LotAreaQuantileBin, dtype: int64)

## wow just side note, this is very cool and extremley fun! I would love it if in class we had more time to actually do the inclass practice so we can ask questions.